include/deal.II/numerics/nonlinear.h

```diff
@@ -0,0 +1,665 @@
+ *  NonlinearSolverSelector<Vector<double>>::AdditionalData::SolverType::kinsol;
+ *
+ * NonlinearSolverSelector<Vector<double>> nonlinear_solver(additional_data);
+ *
+ * // Functions that are required for solving a nonlinear problem,
+ * // which are utilized in both @p KINSOL and @p NOX.
+ * nonlinear_solver.reinit_vector = [&](Vector<double> &x) {...}
+ *
+ * nonlinear_solver.residual =
+ *           [&](const Vector<double> &evaluation_point,
+ *               Vector<double> &      residual) {...}
+ *
+ * nonlinear_solver.setup_jacobian =
+ *           [&](const Vector<double> &current_u,
+ *               const Vector<double> &current_f) {...}
+ *
+ * nonlinear_solver.solve_with_jacobian =
+ *           [&](const Vector<double> &rhs,
+ *               Vector<double> &      dst,
+ *               const double tolerance) {...}
+ *
+ * // Calling the @p solve function with an initial guess.
+ * nonlinear_solver.solve(current_solution);
+ * @endcode
+ */
+template <typename VectorType = Vector<double>>
+class NonlinearSolverSelector
+{
+public:
+  /**
+   * Additional data that will be sent through the NonlinearSolverSelector
+   * class, then into the specific nonlinear solver classes themselves.
+   * Many of these parameters can also be foud in the documentation for
+   * @p KINSOL and @p NOX.
+   */
+  class AdditionalData
+  {
+  public:
+    /**
+     * NonlinearSolverSelector solution strategy. The solver types included in
+     * this class both use a Newton-Krylov solver with a line search and even
+     * Picard iterations (KINSOL).
+     */
+    enum SolutionStrategy
+    {
+      /**
+       * Standard Newton iteration.
+       */
+      newton,
+      /**
+       * Newton iteration with linesearch.
+       */
+      linesearch,
+      /**
+       * Picard iteration.
+       */
+      picard,
+    };
+
+    enum SolverType
+    {
+      /**
+       * Default parameter, will use whatever solver is available
+       * with KINSOL as the priority.
+       */
+      automatic,
+      /**
+       * KINSOL nonlinear solver, part of the SUNDIALS package.
+       */
+      kinsol,
+      /**
+       * NOX nonlinear solver, part of the TRILINOS package.
+       */
+      nox,
+      /*
+       * Use the PETSc SNES solver.
+       */
+      petsc_snes
+    };
+
+    /**
+     * Initialization parameters for NonlinearSolverSelector.
+     *
+     * @param solver_type Nonlinear solver type.
+     * @param strategy Method of solving nonlinear problem, can be 'newton',
+     * 'linesearch', or 'picard'.
+     * @param maximum_non_linear_iterations Maximum number of nonlinear
+     * iterations. This parameter is shared between KINSOL and NOX.
+     * @param function_tolerance Function norm stopping tolerance.
+     * @param relative_tolerance Relative function norm stopping tolerance.
+     * @param step_tolerance Step tolerance for solution update.
+     * @param anderson_subspace_size Anderson acceleration subspace size
```

Comment:
These descriptions do not add anything beyond the name of the variable, and for these variables I can't tell what a parameter represents. Can you expand on this?